sklearn/gaussian_process/kernels.py

```diff
@@ -0,0 +1,532 @@
+            Kernel k(X, Y)
+
+        K_gradient : array (opt.), shape (n_samples_X, n_samples_X, n_params)
+            The gradient of the kernel k(X, X) with repect to the
+            hyperparameter of the kernel. Only returned when eval_gradient
+            is True.
+        """
+        if eval_gradient:
+            K1, K1_gradient = self.k1(X, Y, eval_gradient=True)
+            K2, K2_gradient = self.k2(X, Y, eval_gradient=True)
+            return K1 * K2, np.dstack((K1_gradient * K2[:, :, np.newaxis],
+                                       K2_gradient * K1[:, :, np.newaxis]))
+        else:
+            return self.k1(X, Y) * self.k2(X, Y)
+
+    def __repr__(self):
+        return "{0} * {1}".format(self.k1, self.k2)
+
+
+class ConstantKernel(Kernel):
+    """ Constant kernel.
+
+    Can be used as part of a product-kernel where it scales the magnitude of
+    the other factor (kernel) or as part of a sum-kernel, where it modifies
+    the mean of the Gaussian process.
+    """
+
+    def __init__(self, param_space=1.0):
+        self._parse_param_space(param_space)
+
+    @property
+    def params(self):
+        return np.array([self.value])
+
+    @params.setter
+    def params(self, theta):
+        assert len(theta) == 1
+        self.value = theta[0]
+
+    def __call__(self, X, Y=None, eval_gradient=False):
+        """ Return the kernel k(X, Y) and optionally its gradient.
+
+        Parameters
+        ----------
+        X : array, shape (n_samples_X, n_features)
+            Left argument of the returned kernel k(X, Y)
+
+        Y : array, shape (n_samples_Y, n_features), (optional, default=None)
+            Right argument of the returned kernel k(X, Y). If None, k(X, X)
+            if evaluated instead.
+
+        eval_gradient : bool (optional, default=False)
+            Determines whether the gradient with respect to the kernel
+            hyperparameter is determined. Only supported when Y is None.
+
+        Returns
+        -------
+        K : array, shape (n_samples_X, n_samples_Y)
+            Kernel k(X, Y)
+
+        K_gradient : array (opt.), shape (n_samples_X, n_samples_X, n_params)
+            The gradient of the kernel k(X, X) with repect to the
+            hyperparameter of the kernel. Only returned when eval_gradient
+            is True.
+        """
+        if Y is None:
+            Y = X
+        elif eval_gradient:
+            raise ValueError("Gradient can only be evaluated when Y is None.")
+
+        K = self.value * np.ones((X.shape[0], Y.shape[0]))
+        if eval_gradient:
+            return K, np.ones((X.shape[0], X.shape[0], 1))
+        else:
+            return K
+
+    def __repr__(self):
+        return "{0}".format(self.value)
+
+
+class RBF(Kernel):
+    """ Radial-basis function kernel (aka squared-exponential kernel).
+
+    Both isotropic and anisotropic version are supported.
+    """
+
+    def __init__(self, param_space=1.0):
```

Comment:
Having this `_parse_param_space` function in the base class is in general not a bad idea, but I am afraid it leads to some very non descript parameter names. Lower down you name the specific one for RBF `l` (which should probably be a word), but that isn't evident in the `__init__` and may confuse people.

You have found a good way of propagating parameters even through composition of kernels which I find not easy to do. But I am wondering whether it would be possible to have and propagate names of parameters as well using `__` concatenation, just as pipelines and other meta-estimators do. You could keep the parameters stacked in the array that you have already created, but additionally expose them in a `dict` with appropriate keys.

Don't do this for now, it may actually be more messy than expected. I am just noting stuff for brain storming.